src/Microsoft.AspNetCore.Mvc.TagHelpers/Cache/HybridHtmlFragmentCache.cs

```diff
@@ -0,0 +1,218 @@
+﻿using System;
+using System.IO;
+using System.Text;
+using System.Text.Encodings.Web;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Html;
+using Microsoft.Extensions.Caching.Distributed;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Mvc.TagHelpers.Cache
+{
+    /// <summary>
+    /// This implementation of <see cref="IHtmlFragmentCache"/> is able to switch from 
+    /// <see cref="IMemoryCache"/> to <see cref="IDistributedCache"/> registered
+    /// services using the <code>options-distributed</code> attribute.
+    /// </summary>
+    public class HybridHtmlFragmentCache : IHtmlFragmentCache
+    {
+        private const string PriorityOptionName = "priority";
+        private const string DistributedOptionName = "distributed";
+
+        private readonly IMemoryCache _memoryCache;
+        private readonly IServiceProvider _serviceProvider;
+
+        public HybridHtmlFragmentCache(IMemoryCache memoryCache, IServiceProvider serviceProvider)
+        {
+            _serviceProvider = serviceProvider;
+            _memoryCache = memoryCache;
+        }
+
+        public Task<IHtmlContent> SetAsync(string key, Func<Task<IHtmlContent>> renderContent, HtmlFragmentCacheContext context)
+        {
+            IDistributedCache distributedCache;
+
+            // is the distributed option set?
+            if (EnsureDistributedService(context, out distributedCache))
+            {
+                return SetDistributedAsync(key, renderContent, context, distributedCache);
+            }
+            else
+            {
+                return SetMemoryAsync(key, renderContent, context);
+            }
+        }
+
+        private async Task<IHtmlContent> SetMemoryAsync(string key, Func<Task<IHtmlContent>> renderContent, HtmlFragmentCacheContext context)
+        {
+            var options = new MemoryCacheEntryOptions();
+
+            if (context.ExpiresOn != null)
+            {
+                options.SetAbsoluteExpiration(context.ExpiresOn.Value);
+            }
+
+            if (context.ExpiresAfter != null)
+            {
+                options.SetAbsoluteExpiration(context.ExpiresAfter.Value);
+            }
+
+            if (context.ExpiresSliding != null)
+            {
+                options.SetSlidingExpiration(context.ExpiresSliding.Value);
+            }
+
+            if (context.HasOptions && context.Options.ContainsKey(PriorityOptionName))
+            {
+                CacheItemPriority priority;
+                if (Enum.TryParse(context.Options[PriorityOptionName], out priority))
+                {
+                    options.SetPriority(priority);
+                }
+            }
+
+            // Create an entry link scope and flow it so that any tokens related to the cache entries
+            // created within this scope get copied to this scope.
+            using (var link = _memoryCache.CreateLinkingScope())
```

Comment:
Would the user have an expectation that triggers created in inner contexts cause the outer cache to expire? Would documentation suffice to explain that this wouldn't work for IDistributedCache?